include/pybind11/numpy.h

```diff
@@ -868,6 +908,15 @@ struct format_descriptor<T, detail::enable_if_t<std::is_enum<T>::value>> {
     }
 };
 
+template <typename T>
+struct format_descriptor<T, detail::enable_if_t<detail::array_info<T>::is_array>> {
+    static std::string format() {
+        using detail::_;
+        return (_("(") + detail::array_info<T>::extents() + _(")")).text()
+            + format_descriptor<detail::remove_all_extents_t<T>>::format();
```

Comment:
This is causing the missing symbols with clang/C++14. The `descr` part must be made explicitly `constexpr` to avoid needing a definition for `digits` (i.e. `PYBIND11_DESCR` in place of `constexpr auto` for compatiblity with C++11). The following lines along with removing the `digits` definition reduces binary size slightly for me:

```c++
PYBIND11_DESCR name = _("(") + detail::array_info<T>::extents() + _(")");
return name.text() + format_descriptor<detail::remove_all_extents_t<T>>::format();
```